### PR TITLE
First commit

### PR DESCRIPTION
**Pull request by mcamou**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+name = "events-to-db"
+version = "0.0.1"
+authors = ["Actyx AG <developer@actyx.io>"]
+edition = "2018"
+
+[dependencies]
+actyxos_sdk = { version = "~0.5", features = ["client"] }
+anyhow = "~1.0"
+futures = "~0.3"
+tokio-postgres = {version="~0.6", features = ["with-serde_json-1"]}
```
Comment (wngr) [inline Cargo.toml]:
```suggestion
tokio-postgres = { version="~0.6", features = ["with-serde_json-1"] }
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,9 @@
 # events-to-db
+
 Copies events to a database
+
+## TODO
+
+- Add tests
```
Comment (wngr) [inline README.md]: :trollface:

File: src/main.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    about = "Insert events into a Postgres database.\nEnvironment variables:\n  PGPASSWORD: PostgreSQL password, if not provided with --password/-p\n  AX_EVENT_SERVICE_URI: URL to connect to the Event Service. Default: http://localhost:4454/api/"
+)]
+struct Opt {
+    #[structopt(long, short)]
+    from_start: bool,
+    #[structopt(
+        long,
+        short = "w",
+        about = "If not provided, read from the PGPASSWORD environment variable"
+    )]
```
Comment (wngr) [inline src/main.rs]:
```suggestion
    #[structopt(
        long,
        short = "w",
        env = "PGPASSWORD",
        hide_env_values = true
    )]
```
https://docs.rs/structopt/0.3.21/structopt/#environment-variable-fallback

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    #[structopt(long, short, default_value = "localhost")]
+    host: String,
+    #[structopt(long, short, default_value = "5432")]
+    port: u16,
+    #[structopt(long, short)]
+    username: String,
+    #[structopt(long, short, default_value = "events")]
+    table: String,
+    #[structopt(about = "Subscription set to listen to", default_value = "[{}]")]
+    subscriptions: String,
```
Comment (wngr) [inline src/main.rs]: isn't there a proper type in our sdk?

File: src/main.rs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+pub fn main() -> Result<()> {
```
Comment (wngr) [inline src/main.rs]:
```suggestion
#[tokio::main(flavor = "multi_thread", worker_threads = 2)]
pub async fn main() -> Result<()> {
```

File: src/main.rs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+    let db = db.connect().await?;
+    let offsets = db.get_offsets().await?;
+    debug!("Offset map from database: {:?}", offsets);
+
+    info!(
+        "Database has {} events. Swarm has {} events. Getting {} events to get to the present",
+        &offsets.size(),
+        &present.size(),
+        &present - &offsets
```
Comment (wngr) [inline src/main.rs]: Just curious: That results in an offset map, right? Could you provide a sample print?

File: src/main.rs [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+async fn run_pipeline<C: DbConnection + 'static>(
+    db: Box<dyn Db<C>>,
+    subscriptions: Vec<Subscription>,
+    max_batch_size: usize,
+) -> Result<()> {
+    let event_service = EventService::default();
+    debug!("Connected to EventService");
+
+    let present = event_service.get_offsets().compat().await?;
```
Comment (wngr) [inline src/main.rs]: you could just put the `compat` inside your `main`, and not worry about it here anymore. Using `tokio_02_compat::FutureExt::compat` will enable the future to use both tokio 02 and 03 features, so no need to be frugal

File: src/postgres.rs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+#[async_trait]
+impl Db<PostgresConnection> for Postgres {
+    async fn connect(&self) -> Result<PostgresConnection> {
+        info!("Connecting to database");
+        let conn_str = format!(
+            "host={} port={} user={} password={} dbname={}",
+            self.host, self.port, self.user, self.password, self.db_name
+        );
+        let (client, connection) = tokio_postgres::connect(&conn_str, NoTls).await?;
+        tokio::spawn(Box::pin(connection));
```
Comment (wngr) [inline src/postgres.rs:112]: with tokio03 you could get a `JoinHandle` from this, which you could eventually `.abort()`. But this program probably will just exit if there is any issues with the connection, right?

File: src/main.rs [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    );
+
+    event_service
+        .subscribe_from(offsets, subscriptions)
+        .compat()
+        .await?
+        .map(|e| -> DbEvent { e.into() })
+        .chunks_timeout(max_batch_size, Duration::new(1, 0))
+        .for_each(|chunk| db.insert(chunk).map(|x| x.unwrap()))
+        .await;
```
Comment (wngr) [inline src/main.rs:108]:
```suggestion
        .try_for_each(|chunk| db.insert(chunk))
        .await?;
```

**Review by wngr — COMMENTED** [review]
Mr. Camou on :fire:

File: src/main.rs (hunk 4 above)
Comment (wngr) [reply src/main.rs]: I see, you should be able to directly use `Vec<Subscriptions>` here, maybe you need to provide a custom `from_str`

File: src/main.rs (hunk 6 above)
Comment (mcamou) [reply src/main.rs]: looking at the source, it results in `&present.size() - &offsets.size()`

File: src/postgres.rs (hunk 8 above)
Comment (mcamou) [reply src/postgres.rs:112]: yeah, that's why we have `unwrap()` here: https://github.com/Actyx/events-to-db/pull/1/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR109

File: src/main.rs (hunk 3 above)
Comment (mcamou) [reply src/main.rs]: thanks!

File: src/main.rs (hunk 7 above)
Comment (mcamou) [reply src/main.rs]: nice!

File: src/main.rs (hunk 9 above)
Comment (mcamou) [reply src/main.rs:108]: Hmmm.... does not compile, some incompatibility between `chunks_timeout` and `TryStream`

**Comment by mcamou** [comment]
This is going really fast (~500 records/sec inserting when catching up). Now, why is it using up 2.2G RSS on my machine?

File: src/main.rs (hunk 4 above)
Comment (mcamou) [reply src/main.rs]: Using `parse` from `structopt` worked, although I had to create an auxiliary function.

File: src/main.rs (hunk 6 above)
Comment (mcamou) [reply src/main.rs]: I've removed it anyway, since it does not take the subscriptions into account.

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+# 1c: Build the binary using the actual source code
+COPY src ./src
+RUN touch src/* && \
+  PKG_CONFIG_ALLOW_CROSS=1 OPENSSL_STATIC=true OPENSSL_DIR=/musl cargo build --release --target x86_64-unknown-linux-musl
+
+# 2: Copy the binary  to an empty Docker image
+FROM scratch
+COPY --from=build /usr/src/app/target/x86_64-unknown-linux-musl/release/events-to-db .
+USER 1000
+ENTRYPOINT ["./events-to-db"]
```
Comment (rkuhn) [inline Dockerfile:35]: looking good, way to go!

File: src/main.rs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        opt.table,
+    );
+
+    run_pipeline(
+        Box::new(pg),
+        opt.subscriptions,
+        opt.max_batch_records,
+        opt.max_batch_seconds,
+    )
+    .compat()
```
Comment (rkuhn) [inline src/main.rs:77]: why is this needed?

File: src/postgres.rs [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+            lamports: Vec::with_capacity(capacity),
+            offsets: Vec::with_capacity(capacity),
+            timestamps: Vec::with_capacity(capacity),
+            payloads: Vec::with_capacity(capacity),
+        }
+    }
+}
+
+impl<'a> From<&'a Vec<Event<Payload>>> for DbEventVec<'a> {
+    fn from(events: &'a Vec<Event<Payload>>) -> DbEventVec<'a> {
```
Comment (rkuhn) [inline src/postgres.rs]: might be just the right thing here — in more general cases I’d take a `&'a[Event<Payload>]`, which can easily be obtained from a `v: Vec` using `&*v`.

File: src/postgres.rs [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+        let mut rows = DbEventVec::empty(events.len());
+        for e in events {
+            rows.sources.push(e.stream.source.as_str());
+            rows.semantics.push(e.stream.semantics.as_str());
+            rows.names.push(e.stream.name.as_str());
+            rows.lamports.push(e.lamport.as_i64());
+            rows.offsets.push((e.offset - Offset::ZERO) as i64);
+            rows.timestamps.push(e.timestamp.as_i64());
+
+            // TODO What should we do if extract() fails? It should NEVER fail as long as the payload is JSON, but...
```
Comment (rkuhn) [inline src/postgres.rs]: I’d prefer to keep the process running, skipping the broken event and logging an error.

File: src/postgres.rs [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+#[async_trait]
+impl DbConnection for PostgresConnection {
+    async fn insert(&self, items: Vec<Event<Payload>>) -> Result<()> {
+        let start = Instant::now();
+
+        let num_rows = items.len();
+        debug!("Preparing {} events", num_rows);
+        let rows: DbEventVec = (&items).into();
```
Comment (rkuhn) [inline src/postgres.rs]:
```suggestion
        let rows = DbEventVec::from(&*items);
```
(would be my aesthetic preference)

File: src/postgres.rs [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+            sources,
+        );
+
+        Ok(())
+    }
+
+    async fn get_offsets(&self) -> Result<OffsetMap> {
+        info!("Querying initial offset map");
+        let sql = format!(
+            "SELECT source, MAX(psn) FROM {} GROUP BY source",
```
Comment (rkuhn) [inline src/postgres.rs:215]: This will work nicely for the BAG use-case, one downside is that insertion progress is only reported whenever there is something to insert — skipping over lots of uninteresting events will not be reflected (as opposed to the approach with a separate table).

**Review by rkuhn — APPROVED** [review]
Yes, looking great! Just a few nitpicks.

File: src/main.rs (hunk 11 above)
Comment (mcamou) [reply src/main.rs:77]: Tokio 0.3 / 1.0. Before this it was: [`2be0853` (#1)](https://github.com/Actyx/events-to-db/pull/1/commits/2be085349ff7af27cac2b16172eb21152694ad58#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR74)

File: src/postgres.rs (hunk 15 above)
Comment (mcamou) [reply src/postgres.rs:215]: That is true (I noticed it and removed part of the misleading message). I'd rather keep it simple for now and perhaps make another PR with that change and also perhaps a change to the table schema to reflect our new terminology (and also figure out how to handle tags in EventService v2)

File: src/postgres.rs (hunk 12 above)
Comment (mcamou) [reply src/postgres.rs]: cool... fixed

File: src/postgres.rs (hunk 13 above)
Comment (mcamou) [reply src/postgres.rs]: :+1:

File: src/postgres.rs (hunk 14 above)
Comment (mcamou) [reply src/postgres.rs]: :+1:
